File: 912. Sort an Array/912. Sort an Array.cpp
class Solution {
public:

    void solve(int ind,vector<int>& nums, int m){
        int maxx = ind;
        int l = 2*ind+1;
        int r = 2*ind+2;
        if(l<m && nums[l]>nums[maxx]){maxx=l;}
        if(r<m && nums[r]>nums[maxx]){maxx=r;}
        if(maxx!=ind){swap(nums[ind],nums[maxx]);solve(maxx,nums,m);}
    }

    vector<int> sortArray(vector<int>& nums) {
        const int m = nums.size();
        for(int i=m/2-1;i>=0;--i){solve(i,nums,m);}
        for(int i=m-1;i>=0;--i){swap(nums[i],nums[0]);solve(0,nums,i);}
        return nums;
    }
};

File: 912. Sort an Array/912. Sort an Array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) <= 1:
            return nums
        mid = len(nums) // 2
        left_half = self.sortArray(nums[:mid])
        right_half = self.sortArray(nums[mid:])
        return self.merge(left_half, right_half)
    
    def merge(self, left, right):
        sorted_array = []
        left_index, right_index = 0, 0
        
        while left_index < len(left) and right_index < len(right):
            if left[left_index] < right[right_index]:
                sorted_array.append(left[left_index])
                left_index += 1
            else:
                sorted_array.append(right[right_index])
                right_index += 1
        
        while left_index < len(left):
            sorted_array.append(left[left_index])
            left_index += 1
        
        while right_index < len(right):
            sorted_array.append(right[right_index])
            right_index += 1
        
        return sorted_array

    def medianOfThree(self, nums, low, mid, high):
        if (nums[low] < nums[mid] and nums[mid] < nums[high]) or (nums[high] < nums[mid] and nums[mid] < nums[low]):
            return mid
        elif (nums[mid] < nums[low] and nums[low] < nums[high]) or (nums[high] < nums[low] and nums[low] < nums[mid]):
            return low
        else:
            return high
        